### PR TITLE
154/prevent signin when signed in

### LINKED ISSUE #154: Prevent signIn/Up if already signed in (session exists)
Currently, if the previous user did not explicitly sign out, another existing user can sign in, and will inherit the previous user’s local data. It will become part of the second user’s data, and will be synced to the server and other devices. This is Not Exactly Ideal™ :D 

### Start a hoodie server

In an empty directory, `$ npm init -y && npm i --save hoodie`. In the `package.json`, add 
```javascript
  "hoodie": {
    "inMemory": true
  }
```
Start the server with `$ npm start`, it should run on `http://localhost:8000`.

### Start the client

Make a new html file in a different directory, and copy the hoodie client file into that same directory. Get the client from the `node_modules` folder of the server you just started in the previous step, you’ll find it here: `node_modules/@hoodie/client/dist/hoodie.js`. Load it via script tag. Also load PouchDB, like so:

```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <title>Bugsies</title>
</head>
<body>
  <script src="/hoodie.js"></script>
  <script src="https://cdn.jsdelivr.net/pouchdb/6.2.0/pouchdb.min.js"></script>
</body>
</html>
```
Go ahead and load this `index.html` in __AN INCOGNITO WINDOW__ in your browser, for example via OS X's included python server: `$ python -m SimpleHTTPServer`. This will make it available at `http://localhost:8080`.

## Triggering the bug

We'll need to create two users to demonstrate the bug, so let’s do that. Run through these steps one by one in the browser's dev console. 

```javascript
var hoodie = new Hoodie({
  url: 'http://localhost:8080',
  PouchDB: PouchDB
})
```
```javascript
hoodie.account.signUp({username: 'firstuser', password: '123'}).then(dir)
```
```javascript
hoodie.account.signIn({username: 'firstuser', password: '123'}).then(dir)
```
```javascript
hoodie.store.add({author:'firstuser', message:'Hellooo!'}).then(dir)
```
```javascript
hoodie.store.findAll().then(dir)
```
The last command should log an array with a single object, namely the one we added in the second-to-last command. Now log out with `hoodie.account.signOut()`.

Now create the second user:
```javascript
hoodie.account.signUp({username: 'seconduser', password: '123'}).then(dir)
```
```javascript
hoodie.account.signIn({username: 'seconduser', password: '123'}).then(dir)
```
```javascript
hoodie.store.add({author:'seconduser', message:'Ciao!'}).then(dir)
```
```javascript
hoodie.store.findAll().then(dir)
```
Again, the last command should log an array with a single object, namely the one we added in the second-to-last command. Now __DON'T__ log out, instead sign in with the first user:
```javascript
hoodie.account.signIn({username: 'firstuser', password: '123'}).then(dir)
```
This succeeds. When we `hoodie.store.findAll().then(dir)` now, it will return an array with two objects, one from `firstuser` and one from `seconduser`.

## Proposed fix:

Ideally, we would completely prevent `signIn` while there is a user session. It’s also possible to `signUp` while already signed in, not sure whether that also needs to be prevented.

### PR DESCRIPTION
**Pull request by espy**
Closes #154 by checking on signIn whether a session exists, and whether that session belongs to anyone but the user trying to sign in. :no_good_man: 

Oh, I didn’t touch `signUp` for now, although it’s mentioned in the issue. 

Ready for review @hoodiehq/maintainers !

### CODE REVIEW

File: test/integration/sign-in-test.js [hunk 1]
```diff
@@ -85,3 +85,36 @@ test('sign in', function (t) {
 
   .catch(t.error)
 })
+
+test('sign in while signed in fails', function (t) {
```
Comment (mbad0la) [inline test/integration/sign-in-test.js]: The test name seems a bit ambiguous. Wouldn't `prevent sign-in when already signed in` be better?

File: test/integration/sign-in-test.js [hunk 2]
```diff
@@ -85,3 +85,36 @@ test('sign in', function (t) {
...
+  })
+
+  nock(baseURL)
+    .put('/session')
+    .reply(201, signInResponse)
+
+  account.signIn(options)
+
+  .then(function (signInResult) {
+    t.pass('signs in')
```
Comment (mbad0la) [inline test/integration/sign-in-test.js]: Isn't this check redundant?

File: test/integration/sign-in-test.js (hunk 1 above)
Comment (espy) [inline test/integration/sign-in-test.js]: I agree, yours is clearer. Will change.

File: test/integration/sign-in-test.js (hunk 2 above)
Comment (espy) [inline test/integration/sign-in-test.js]: Also a good find!

**Comment by espy** [comment]
Thanks for those, @mbad0la. As for the `signUp`, I’ll leave it out, we might have situations later where one user can sign up other users, plus it doesn’t break anything if we allow this.

**Comment by gr2m** [comment]
I’d suggest when the last typo is fixed, use the `Squash and merge` button with the commit message "fix: prevent signIn if session present". 

This is just fyi, I know it’s rather complex: the last 3 commit messages should have used `test: ` as they only touch test files. All commit messages with a `fix: ` commit will appear in the change logs for the users of this library, and for them "more descriptive test name" or "typo" is not very helpful. Does that make sense?

Great review @mbad0la 👍

**Comment by mbad0la** [comment]
> This is just fyi, I know it’s rather complex: the last 3 commit messages should have used test: as they only touch test files. All commit messages with a fix: commit will appear in the change logs for the users of this library, and for them "more descriptive test name" or "typo" is not very helpful.

I was confused about this myself. Thanks for stepping in @gr2m 

> fix: prevent signIn if session present"

Is this accurate? The fix will prevent the propagation of the `signin` event, but not the `signIn` method when `cache.username === options.username`

**Comment by gr2m** [comment]
> The fix will prevent the propagation of the signin event, but not the signIn method when cache.username === options.username

I’m not sure what you mean with it will not prevent the sign in method? What happens is that if the curernt user already has a session and tries to sign in with a different username than the current who they are signed in with, then don’t do anything (no events, no change of state) and reject with an explanatory error message

**Comment by espy** [comment]
I think this is good now, are you happy with the changes, @mbad0la ? If you approve the changes, I can squash and merge :) 

Thanks again!

**Comment by mbad0la** [comment]
LGTM @espy :+1:

**Pull request merged**